### PR TITLE
adding Jeti EX Bus Receiver

### PR DESCRIPTION
**Pull request by marv-t**
Add the Jeti EX Bus receiver.
The telemetry is not supported now, only the channels are processed.

### CODE REVIEW

File: src/main/rx/jetiexbus.c [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+	return ret_val; 
+} 
+
+// Jeti Ex Bus CRC calculations
+uint16_t calcCRC16(volatile uint8_t *pt, uint8_t msgLen)
+{
+    uint16_t crc16_data=0;
+   	uint8_t mlen=0;
+	
+    while(mlen<=msgLen) {
```
Comment (ledvinap) [inline src/main/rx/jetiexbus.c]: crc of msgLen+1 bytes is computed ... this is a bit confusing ..

File: src/main/rx/jetiexbus.c [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+			if((jetiExBusFrame[2]<=JETIEXBUS_MAX_FRAME_SIZE) && (jetiExBusFrame[2]>=JETIEXBUS_MIN_FRAME_SIZE)){
+				jetiExBusFrameLength = jetiExBusFrame[2];
+			}else{
+				jetiExBusFrameReset();
+			}
+		}
+    	
+    // Done?
+    if (jetiExBusFrameLength == jetiExBusFramePosition) {
+		crc = ((uint16_t)jetiExBusFrame[jetiExBusFramePosition - 1] << 8) + (jetiExBusFrame[jetiExBusFramePosition - 2]);
```
Comment (ledvinap) [inline src/main/rx/jetiexbus.c]: CRC16 of message including (senders) crc will be same for all valid messages, and probably 0 in this case.. There is no need to check explicitly

File: src/main/rx/jetiexbus.c (hunk 1 above)
Comment (marv-t) [inline src/main/rx/jetiexbus.c]: Ok, i see this is a little bit confuse. I will correct it.

File: src/main/rx/jetiexbus.c (hunk 2 above)
Comment (marv-t) [inline src/main/rx/jetiexbus.c]: You're right , I 'm going to change the same.

**Comment by ledvinap** [comment]
One possible problem is that frame checking and decoding is called from USART receive interrupt. This interrupt is quite high priority and it may break other code (softserial is one example).
It would be better to check and parse frame in mail loop, possibly in `jetiExBusFrameStatus`

File: src/main/rx/jetiexbus.c [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+	data ^= (uint8_t)(crc) & (uint8_t)(0xFF);   
+	data ^= data << 4;   
+	ret_val = ((((uint16_t)data << 8) | ((crc & 0xFF00) >> 8))             
+			  ^ (uint8_t)(data >> 4)
+			  ^ ((uint16_t)data << 3));   
+	return ret_val; 
+} 
+
+// Jeti Ex Bus CRC calculations
+uint16_t calcCRC16(volatile uint8_t *pt, uint8_t msgLen)
```
Comment (ledvinap) [inline src/main/rx/jetiexbus.c]: No `volatile` here

File: src/main/rx/jetiexbus.c [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+    uint32_t now;
+    static uint32_t jetiExBusTimeLast, jetiExBusTimeInterval;
+	uint16_t crc = 0;
+	
+    // Check if we shall reset frame position due to time
+    now = micros();
+    jetiExBusTimeInterval = now - jetiExBusTimeLast;
+    jetiExBusTimeLast = now;
+	
+    if (jetiExBusTimeInterval > JETIEXBUS_MAX_FRAME_TIME) {
```
Comment (ledvinap) [inline src/main/rx/jetiexbus.c]: This is actually interbyte timeout. It may be shorter than frame time

File: src/main/rx/jetiexbus.c [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+
+    // Only do this if we are receiving to a frame
+    if (jetiExBusDataIncoming == true) {
+		// Store in frame copy
+        jetiExBusFrame[jetiExBusFramePosition] = (uint8_t)c;
+        jetiExBusFramePosition++;
+    }
+	
+	// Check the header for the messagelength
+	if((jetiExBusFramePosition==JETIEXBUS_HEADER_LEN) && (jetiExBusHeaderReady==false)) {
```
Comment (ledvinap) [inline src/main/rx/jetiexbus.c]: The headerReady flag is redundant, `(jetiExBusFramePosition==JETIEXBUS_HEADER_LEN)` is true only once with current code

File: src/main/rx/jetiexbus.c (hunk 4 above)
Comment (marv-t) [inline src/main/rx/jetiexbus.c]: This timeout is meant for frames that have been stopped or started somewhere.

File: src/main/rx/jetiexbus.c (hunk 5 above)
Comment (marv-t) [inline src/main/rx/jetiexbus.c]: Yes, the flag is obsolete.

File: src/main/rx/jetiexbus.c (hunk 3 above)
Comment (marv-t) [inline src/main/rx/jetiexbus.c]: removed...

File: src/main/rx/jetiexbus.c (hunk 4 above)
Comment (ledvinap) [inline src/main/rx/jetiexbus.c]: Yes. But time is counted from last received byte, not from start of last frame (as it was in sbus code).

Normally bytes arrive about 80us  apart (125kbd), `JETIEXBUS_MAX_FRAME_TIME` may be as low as 160us (last received byte is overwritten after this  time), but personallly I'll use ~500us as safety margin. Using  JETIEXBUS_MAX_FRAME_TIME longer than interframe delay may theoretically result in wrong sync -  start-of-frame byte may be inside packet and decoder may sync to it (SOF bytes may be less than header length apart)

Comment (marv-t) [inline src/main/rx/jetiexbus.c]: The #define I'll renamed from JETIEXBUS_MAX_FRAME_TIME to JETIEXBUS_MIN_FRAME_GAP.

**Comment by marv-t** [comment]
Concerning the ISR:
The CRC test I will leave in the ISR, since you have to respond very quickly for telemetry.
The channel data is, as suggested, extracted in jetiExBusFrameStatus.

Concerning the telemetry:
In future, the telemetry data is also transmitted. Currently this is not so easy, because there is no shared port for receiver and telemetry.

The Jeti telemtetrie is a half-duplex protocol. The TX and RX lines are connected with a resistor. The master sends a telemetry request, then we have 4ms time to post a response. The Master released the TX line at this time - equivalent as HOTT.

I am going to implement the telemetry protocol. How do I transfer the Port of telemetry routines I have to even look at. As displayed in the GUI or as soon as the Jeti EX bus is selected, the telemetry is also activated, I still don't know.
But as the time window is very small, the call for sending the data should be very timely.

@Petr
Thanks for the quick replies!

File: src/main/rx/jetiexbus.c (hunk 4 above)
Comment (ledvinap) [inline src/main/rx/jetiexbus.c]: Thanks

**Comment by ledvinap** [comment]
@marv-t : As I read the doc, the 4ms deadline is for first byte of reply, not whole reply frame. And even 
if whole frame must fit, there is 800us available (4ms - 40B * 125000baud)
 4ms is huge amount of time in CleanFlight ...

Take a look at smartport telemetry: https://github.com/cleanflight/cleanflight/blob/master/src/main/telemetry/smartport.c#L298

Frames are received in interrupt, but rest of processing is called from main loop. Reply delay is short enough. This will be a bit more complicated for receiver, RC processing is called at lower frequency.

It may be better to implement 'telemetry capable' receiver concept in CF instead of splitting code into receiver and telemetry part.

**Comment by marv-t** [comment]
The telemetry capable receivers was my first thought, as this is logical. But in CF everything is well structured and parted, so I thought that this goes against the guidelines.

The 4ms you're right, this is for the first byte. I overlooked the "Releases the line for at least 4ms".
But if this is allowed, the telemetry routines are inserted in jetexbus.c.

The update rate of the telemetry is defined by the loop ().
https://github.com/cleanflight/cleanflight/blob/master/src/main/telemetry/telemetry.c#L83

I think that's fast enough ;-)

The receiver must then be included in telemtry.c.

Ok, enough to do ...

**Comment by ledvinap** [comment]
IMO it's wrong to implement JETI as telemetry module with receiver capability. The system expects single receiver but allows multiple telemetry interfaces. Also receiver code is be mission-critical, telemetry problems won't have that severe consequences .. 

Maybe some stub in `telemetry\` that will call into RX code?

@hydra is in charge, I am just stating my personal opinion here..

**Comment by marv-t** [comment]
I think I I've expressed myself wrong. Sorry my english is not the best.

I meant that the calls for the functions in the telemetry\ are consistent.
The code will be in rx\, telemetry functions are only diverted.

**Comment by hydra** [comment]
Can you rebase this against latest master please

**Comment by marv-t** [comment]
Is made this evening.

**Comment by marv-t** [comment]
I had problems with the merge from master. I'll create a new pull request with a new branch.

**Comment by hydra** [comment]
@marv-t Hiya, any update on this?

**Comment by marv-t** [comment]
Hi Dominic,

I'm in contact with Jeti to implement the telemtry. I have some issues with
the updatespeed, but i think its a small bug.
I was on holidays the last 2 weeks. Tonight i work on the implementation...

Have a nice day,
Thomasq

Von meinem iPad gesendet

Am 18.01.2016 um 13:29 schrieb Dominic Clifton <notifications@github.com>:

@marv-t <https://github.com/marv-t> Hiya, any update on this?

—
Reply to this email directly or view it on GitHub
<https://github.com/cleanflight/cleanflight/pull/1563#issuecomment-172516228>
.

**Comment by marv-t** [comment]
This is the new PR: #1709

**Pull request closed without merging**